### ISSUE TITLE
sprout can't talk to chronos in an IPv6 environment

### ISSUE THREAD

**Issue by dtheobald** [Metaswitch/sprout#1510]
#### Symptoms

Sprout issues SPROUT_CHRONOS_COMM_ERROR alarm when trying to communicate with chronos on an IPv6 deployment. The following entry is also seen in sprout's log:
```
UTC Error httpconnection.cpp:643: http://127.0.0.1:7253/timers failed at server 127.0.0.1 : Couldn't connect to server (7) : fatal
```
Chronos is in fact listening on IPv6 localhost @ port 7253:
```
tcp6       0      0 ::1:7253                :::*                    LISTEN      3693/chronos    
tcp6       0      0 2620:104:400c:112d:7253 :::*                    LISTEN      3693/chronos    
```

#### Impact

Timers don't work

#### Release and environment 

release-103

#### Steps to reproduce
1) Deploy an IPv6 system
2) Make a call

Diags are available

**Comment by mirw** [comment]
How is your IPv6 system deployed?  I think when we've tested this before, we've used IPv6 on all external interfaces, but IPv4 on localhost interfaces (in particular for Chronos).  Does that work?

(Not saying that we shouldn't change this, but it would be useful to understand how IPv6 is being configured.)

**Comment by dtheobald** [comment]
The system is configured with IPv4 only on the management interface and IPv6 only on the signaling interface (i.e., single stack on both interfaces). Perhaps PC testing was done on a multiple stack system where (I believe) using 127.0.0.1 would work. As an experiment in a way to correct this problem, I made the following change to sprout:
```
@@ -1881,7 +1881,7 @@ int main(int argc, char* argv[])
   {
     // Create a connection to Chronos.
     std::string port_str = std::to_string(opt.http_port);
-    std::string chronos_callback_host = "127.0.0.1:" + port_str;
+    std::string chronos_callback_host = "localhost:" + port_str;
 
     // We want Chronos to call back to its local sprout instance so that we can
     // handle Sprouts failing without missing timers.
@@ -1890,7 +1890,7 @@ int main(int argc, char* argv[])
       chronos_callback_host = "[::1]:" + port_str;
     }
 
-    std::string chronos_service = "127.0.0.1:7253";
+    std::string chronos_service = "localhost:7253";
     TRC_STATUS("Creating connection to Chronos %s using %s as the callback URI",
                chronos_service.c_str(),
                chronos_callback_host.c_str());
```
Not surprisingly, using ```localhost``` instead of the IPv4 127.0.0.1 IP address corrects the problem. However, this draws into question the many other uses of 127.0.0.1 in the source code.

**Comment by mirw** [comment]
Is it a requirement to run without even an IPv4 loopback interface?  As you say, there are a number of references to 127.0.0.1, and the simpler approach here is to create an IPv4 loopback interface (but no other IPv4 interfaces) - this should be invisible to the outside world.

**Comment by mirw** [comment]
BTW, I'm nervous about switching to using "localhost" throughout.  The reason we actually use `127.0.0.1` here is down to IPv6 - the default behavior on an IPv6-enabled system is/was to attempt DNS resolution using AAAA, and only after that fails, fall back to A.  If `localhost` isn't configured locally to map to `::1` (and it's not configured as such by default - have you changed your `/etc/hosts` file?) then the system sends a AAAA DNS query externally for localhost.  Ideally, this fails cleanly, meaning that we retry with an A record query, succeed quickly, and just have a slightly delayed DNS resolution.  However, we have seen instances in which the external DNS server returns a response that causes the local system to believe that `localhost` is not resolvable anywhere, even as an A record - in that case, you can't actually ping localhost.  If we need to go this route, I fear we'll need to make the localhost hostname configurable.

**Comment by dtheobald** [comment]
I've taken a closer look at the problem and what I've found may be too much for here, but I'll give it a try:

Sprout not being able to talk to chronos via ```127.0.0.1``` was caused by L3's default setting change in ```/etc/chronos/chronos.conf``` whereby the 'bind-address' was set to the signaling interface's IP address. This change seems innocent enough on the surface, but there are unintended consequences for an IPv6 deployment; that being, that chronos can no longer be contacted locally using the IPv4 ```127.0.0.1``` address since chronos is specifically bound to an IPv6 address instead of its default ```0.0.0.0```. Given this, one might think that the fix for this is to remove the 'bind-address' setting and allow chronos to use its default; which works for IPv4 & 6 localhost communication. Unfortunately, this doesn't allow chronos(s) to talk to each other in an IPv6 deployment, since binding to ```0.0.0.0``` only allows IPv4 connections except for those to IPv6 localhost. Given this, I changed the 'bind-address' to ```::0``` which allows chronos to accept both IPv4 and IPv6 connections. Sounds good, right? By all accounts everything looked good (i.e., no alarms, etc), but being suspicious I set chronos to log level 5 just to make sure and here's what I saw in the log:

```
Debug replicator.cpp:174: http://2620:104:400c:112d::c0a8:a553:7253/timers/001740cc6920000080d2300000400021 failed at server. URL using bad/illegal format or missing URL(3). Fata
```

Chronos is not formatting the URL correctly for IPv6 which doesn't get ENT logged or alarmed (should be another issue). So I found a couple of spots in chronos' source (could be more?) where the address formatting needed to change and made the corrections so that I'd have a functional chronos. Here are those changes:

```
diff --git a/src/chronos_internal_connection.cpp b/src/chronos_internal_connection.cpp
index 1c5189a..16de363 100644
--- a/src/chronos_internal_connection.cpp
+++ b/src/chronos_internal_connection.cpp
@@ -158,6 +158,12 @@ void ChronosInternalConnection::resynchronize()
     }
 
     std::string server_to_sync = address + ":" + std::to_string(port);
+    // Determine if we're IPv4 or IPv6.
+    struct in6_addr dummy_addr;
+    if (inet_pton(AF_INET6, address.c_str(), &dummy_addr) == 1)
+    {
+      server_to_sync = "[" + address + "]:" + std::to_string(port);
+    }
     HTTPCode rc = resynchronise_with_single_node(server_to_sync,
                                                  cluster_nodes,
                                                  localhost);
diff --git a/src/globals.cpp b/src/globals.cpp
index 9e4668f..f6c1781 100644
--- a/src/globals.cpp
+++ b/src/globals.cpp
@@ -58,7 +58,7 @@ Globals::Globals(std::string config_file,
 
   // Describe the configuration file format.
   _desc.add_options()
-    ("http.bind-address", po::value<std::string>()->default_value("0.0.0.0"), "Address to bind the HTTP server to")
+    ("http.bind-address", po::value<std::string>()->default_value("::0"), "Address to bind the HTTP server to")
     ("http.bind-port", po::value<int>()->default_value(7253), "Port to bind the HTTP server to")
     ("cluster.localhost", po::value<std::string>()->default_value("localhost:7253"), "The address of the local host")
     ("cluster.joining", po::value<std::vector<std::string>>()->multitoken()->default_value(std::vector<std::string>(), "HOST"), "The addresses of nodes in the cluster that are joining")
diff --git a/src/timer.cpp b/src/timer.cpp
index de19195..b7d4cdd 100644
--- a/src/timer.cpp
+++ b/src/timer.cpp
@@ -146,7 +146,17 @@ std::string Timer::url(std::string host)
       __globals->get_bind_port(port);
     }
 
-    ss << "http://" << address << ":" << port;
+    // Determine if we're IPv4 or IPv6.
+    struct in6_addr dummy_addr;
+    if (inet_pton(AF_INET6, address.c_str(), &dummy_addr) == 1)
+    {
+      ss << "http://[" << address << "]:" << port;
+    }
+    else
+    {
+      ss << "http://" << address << ":" << port;
+    }
+
   }
 
   ss << "/timers/";
```
**Note**: I didn't use the is_ipv6 function in cpp-common because I was getting a link error that I didn't feel like resolving so I just hacked its' source directly into chronos' code figuring PC would have a better way of fixing it anyway.

I think one of the takeaways here is that any socket binding to ```0.0.0.0``` (the way that chronos was doing it) should be changed to ```::0``` so that the socket will function in both an IPv4 and IPv6 deployment. I made a cursory scan of all the source for such bindings, but I saw nothing obvious. However, PC should do their own search since they're intimately familiar with code.

Also, there's something about using the IPv4 localhost IP address (literal - ```127.0.0.1```) that bothers me, but I have nothing concrete to go on other than a gut (bad) feeling about it. I think it would behove PC to replace all references to ```127.0.0.1``` with ```localhost``` and allow it to be resolved by DNS or APIs as indicated in [RFC6761](https://tools.ietf.org/html/rfc6761#section-6.3) section 6.3 (excerpt below):

6.3.  Domain Name Reservation Considerations for "localhost."


   The domain "localhost." and any names falling within ".localhost."
   are special in the following ways:

   1.  **Users are free to use localhost names as they would any other
       domain names.  Users may assume that IPv4 and IPv6 address
       queries for localhost names will always resolve to the respective
       IP loopback address.**

   2.  Application software MAY recognize localhost names as special, or
       MAY pass them to name resolution APIs as they would for other
       domain names.

   3.  Name resolution APIs and libraries SHOULD recognize localhost
       names as special and SHOULD always return the IP loopback address
       for address queries and negative responses for all other query
       types.  Name resolution APIs SHOULD NOT send queries for
       localhost names to their configured caching DNS server(s).

   4.  Caching DNS servers SHOULD recognize localhost names as special
       and SHOULD NOT attempt to look up NS records for them, or
       otherwise query authoritative DNS servers in an attempt to
       resolve localhost names.  Instead, caching DNS servers SHOULD,
       for all such address queries, generate an immediate positive
       response giving the IP loopback address, and for all other query
       types, generate an immediate negative response.  This is to avoid
       unnecessary load on the root name servers and other name servers.

   5.  Authoritative DNS servers SHOULD recognize localhost names as
       special and handle them as described above for caching DNS
       servers.

   6.  DNS server operators SHOULD be aware that the effective RDATA for
       localhost names is defined by protocol specification and cannot
       be modified by local configuration.

   7.  DNS Registries/Registrars MUST NOT grant requests to register
       localhost names in the normal way to any person or entity.
       Localhost names are defined by protocol specification and fall
       outside the set of names available for allocation by registries/
       registrars.  Attempting to allocate a localhost name as if it
       were a normal DNS domain name will probably not work as desired,
       for reasons 2, 3, 4, and 5 above.

**Comment by mirw** [comment]
Unclaiming.  Please can someone who is looking at IPv6 pick this up?

**Comment by sathiyan-sivathas** [comment]
I'm working on a similar bug so I'll pick this up.
